spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala

```diff
@@ -451,6 +468,23 @@ class CometSparkSessionExtensions
       }
     }
   }
+
+  // CometExec already wraps a `ColumnarToRowExec` for row-based operators. Therefore,
+  // `ColumnarToRowExec` is redundant and can be eliminated.
+  //
+  // It was added during ApplyColumnarRulesAndInsertTransitions' insertTransitions phase when Spark
+  // requests row-based output such as `collect` call. It's correct to add a redundant
+  // `ColumnarToRowExec` for `CometExec`. However, for certain operators such as
+  // `CometCollectLimitExec` which overrides `executeCollect`, the redundant `ColumnarToRowExec`
+  // makes the override ineffective. The purpose of this rule is to eliminate the redundant
+  // `ColumnarToRowExec` for such operators.
+  case class EliminateRedundantColumnarToRow(session: SparkSession) extends Rule[SparkPlan] {
+    override def apply(plan: SparkPlan): SparkPlan = {
+      plan.transform { case ColumnarToRowExec(child: CometCollectLimitExec) =>
+        child
```

Comment:
Hmm, this looks like a bit dangerous if `ColumnarToRowExec` + `CometCollectLimitExec` is not end of the query.

I think the assumption here is the query is to collect data from `ColumnarToRowExec` + `CometCollectLimitExec`. So `executeCollect` is called on `ColumnarToRowExec` which makes ineffective of `CometCollectLimitExec`'s `executeCollect`.

The more correct one maybe:

```scala
plan match {
  case ColumnarToRowExec(child: CometCollectLimitExec) =>
        child
  case other => other
}
```

Reply:
> I'd like to point out that `ColumnarToRowExec + CometCollectLimitExec` will always be the end of the query as `CollectLimitExec` is the end of query.

Yes, this is usually the case. But I remember that at some special cases, users can produce a query tree that some others operators on top of `CollectLimitExec`. I think this is why `CollectLimitExec` still implements `doExecute` not just `executeCollect`.